TA4: Enclave restart test

QuentinI opened the pull request:
Closes #<ISSUE_NUMBER>
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
- Adds support for running batcher in TEE to the devnet test helpers
- Uses it to support testing restarting TEE batcher in the batcher restart tests
- Adds a test TestEnclaveRestart
- Skips TestSmokeWithTEE on non-enclave machines
- Add the tests (the new TestEnclaveRestart and the old TestSmokeWithTEE) to CI with initiating a new machine (takes 47min to finish, see an example run [here](https://github.com/EspressoSystems/optimism-espresso-integration/actions/runs/21303586141/job/61326713902))

### This PR does not:
- Make it consistent passing. It passes three times locally for me, but when running in CI it can sometimes fail (2 pass 2 fail), and adjusting to different timeout can lead to other errors as well.
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

### Key places to review:
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

### How to test this PR: 
This test requires a Nitro enclave-enabled EC2 instance because it runs the batcher inside an AWS Nitro enclave. You cannot run this on a regular machine.
Run this on a Nitro enclave-enabled machine for testing:
```
just devnet-TestEnclaveRestart-test
```
If only changes `run-enclave.sh` (which means you don't need redoing build-devnet)
```
ESPRESSO_RUN_ENCLAVE_TESTS=true U_ID=$(id -u) GID=$(id -g) go test -timeout 30m -p 1 -count 1 -v -run TestEnclaveRestart ./espresso/devnet-tests/...
```
And you can see
```
--- PASS: TestEnclaveRestart (1278.92s)
PASS
ok      github.com/ethereum-optimism/optimism/espresso/devnet-tests     1306.740s
```
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->
<!-- [ ] For integration projects, make sure it won't break backward compatibility. -->

<!-- Details on maintaining backward compatibility for integration projects: -->
<!-- * Try to keep changes additive: add new, optional methods, flags, or parameters instead of modifying or removing existing functionality. -->
<!-- * If modification is necessary, it should be either a clear bug fix or guarded by a config/feature flag.  -->
<!-- * Follow Open Closed Principle and Interface Segregation Principle, clients should not be forced to depend on interfaces they do not use. -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210592590143369

QuentinI committed:
Test batcher restart in enclave

dailinsubjam reviewed (approved):
Looks good.

philippecamacho commented:
Why is it possible to run the test locally even without a TEE?

dailinsubjam commented:
@philippecamacho If you're not on a enclave-enabled instance, you'll hit [this line](https://github.com/EspressoSystems/optimism-espresso-integration/blob/celo-integration-rebase-13.2/op-batcher/batcher/service.go#L162) and it will automatically skip enclave-related operations.

dailinsubjam commented on espresso/devnet-tests/batcher_restart_test.go:
```diff
@@ -6,13 +6,26 @@ import (
 )
 
 func TestBatcherRestart(t *testing.T) {
+	testRestart(t, false)
+}
```
I just realized to restart op-batcher-tee we not only need profile to be tee but also need things like restarting the service op-batcher-tee specifically.

QuentinI replied:
Oh, that's a big catch for sure, thanks. No idea how it passed for me in the first place.

philippecamacho commented:
> @philippecamacho If you're not on a enclave-enabled instance, you'll hit [this line](https://github.com/EspressoSystems/optimism-espresso-integration/blob/celo-integration-rebase-13.2/op-batcher/batcher/service.go#L162) and it will automatically skip enclave-related operations.

I see, but when this test runs in CI, is the batcher executed inside the TEE?

QuentinI commented:
> I see, but when this test runs in CI, is the batcher executed inside the TEE?

No, it isn't. Only non-tee test runs.

QuentinI commented:
@dailinsubjam @philippecamacho this works now, I've added a `devnet-enclave-tests` target to justfile which you can run on an AWS Nitro instance to verify (don't forget you need to be in nix shell).

Not set up to run in CI unfortunately, building dockers on AWS machine takes too long and GitHub kills the action. To run this in CI, we need to set up building dockers in one action and re-using in subsequent ones, including uploading them to the AWS Nitro instance, which I'd argue out of scope for this PR, LMK if you disagree.

dailinsubjam commented on espresso/devnet-tests/devnet_tools.go.orig:
I see we’ve added a backup of the backup of the original… bold move 😎

QuentinI replied:
Oh heck, this is from mergiraf, I'll remove it stat 🤣

dailinsubjam commented on espresso/devnet-tests/devnet_tools.go:
```diff
@@ -153,50 +253,52 @@ func (d *Devnet) Up() (err error) {
+		"docker", "compose", "--profile", d.getProfile(), "down", serviceName,
 	)
 	return cmd.Run()
 }
```
This reminds me of one thing: `op-batcher-tee` container spawns an enclave container (`batcher-enclaver-xxx`) that doesn’t stop automatically when `op-batcher-tee` stops. In Docker Compose I’ve been cleaning it up manually via `espresso/scripts/shutdown.sh`, but that’s probably not the best approach.
I'm thinking we may see the same issue here, even if `op-batcher-tee` exits, `batcher-enclaver-xxx` may keep running. 
One possible way is to add a cleanup hook when shutting down`op-batcher-tee` (not sure whether it's supported) or also add a manual shutdown here. WDYT?

QuentinI replied:
Oh interesting. I'll see what we can do here

dailinsubjam reviewed (dismissed):
`TestEnclaveRestart` pass for me.
But `docker ps` after the test, I can see
```
docker ps
CONTAINER ID   IMAGE                    COMMAND                  CREATED          STATUS          PORTS     NAMES
224fe2454142   op-proposer:espresso     "/bin/entrypoint.sh …"   13 minutes ago   Up 10 minutes             espresso-op-proposer-1
78fefd376a3b   op-challenger:espresso   "/bin/entrypoint.sh …"   13 minutes ago   Up 10 minutes             espresso-op-challenger-1
099c18b6b26a   op-batcher:espresso      "op-batcher --espres…"   13 minutes ago   Up 10 minutes             espresso-op-batcher-1
```
remains running.
We can open a separate ticket for cleaning them up.

dailinsubjam reviewed (commented):
Oops, I just found `tee` is not triggered automatically though I'm on a AWS Nitro, will take a further look.

QuentinI committed:
Shutdown tee on exit

QuentinI commented:
@dailinsubjam I've mixed up order of operations during rebase, my tee flag was set after the devnet was spun up 🤦 
Fixed it, going to re-run the thing now and see if it fixes the leftover containers (it should 🤞 )

QuentinI committed:
Wait for devent to start up

dailinsubjam commented:
I got this
```
op-geth-sequencer-1  | INFO [10-16|15:47:27.716] Persisted trie from memory database      nodes=1542 size=170.63KiB time=3.279752ms    gcnodes=5765 gcsiz
e=1.63MiB gctime=11.571326ms livenodes=3207 livesize=978.01KiB
op-geth-sequencer-1  | INFO [10-16|15:47:27.717] Writing cached state to disk             block=366 hash=aaa89c..41bd2d root=5e5706..bcf590
op-geth-sequencer-1  | INFO [10-16|15:47:27.717] Persisted trie from memory database      nodes=25   size=7.76KiB   time="116.902µs"   gcnodes=0    gcsiz
e=0.00B   gctime=0s          livenodes=3182 livesize=970.25KiB
op-geth-sequencer-1  | INFO [10-16|15:47:27.717] Writing cached state to disk             block=240 hash=b04582..df2713 root=67d352..32720e
op-geth-sequencer-1  | INFO [10-16|15:47:27.717] Persisted trie from memory database      nodes=269  size=42.88KiB  time="734.222µs"   gcnodes=0    gcsiz
e=0.00B   gctime=0s          livenodes=2913 livesize=927.37KiB
op-geth-sequencer-1  | INFO [10-16|15:47:27.717] Writing snapshot state to disk           root=1751bc..130f87
op-geth-sequencer-1  | INFO [10-16|15:47:27.717] Persisted trie from memory database      nodes=0    size=0.00B     time="2.68µs"      gcnodes=0    gcsiz
e=0.00B   gctime=0s          livenodes=2913 livesize=927.37KiB
op-geth-sequencer-1  | INFO [10-16|15:47:27.721] Blockchain stopped
l1-geth-1            | INFO [10-16|15:47:28.004] Starting work on payload                 id=0x03b6dac60f0c1813
l1-geth-1            | INFO [10-16|15:47:28.004] Updated payload                          id=0x03b6dac60f0c1813 number=246 hash=e17a7f..15b198 txs=0 with
drawals=0 gas=0          fees=0           root=8c906e..654623 elapsed="132.132µs"
espresso-dev-node-1 exited with code 143
l1-validator-1 exited with code 0
l1-beacon-1          | Oct 16 15:47:28.614 INFO  Shutting down..                               reason: Success("Received SIGTERM") 
l1-beacon-1          | Oct 16 15:47:28.614 INFO  Saved DHT state                              service: "network"
l1-beacon-1          | Oct 16 15:47:28.614 INFO  Network service shutdown                     service: "network"
l1-beacon-1          | Oct 16 15:47:28.618 INFO  Saved beacon chain to disk                   service: "network"
op-geth-sequencer-1 exited with code 0 
l1-beacon-1 exited with code 0
l1-geth-1 exited with code 137
--- FAIL: TestEnclaveRestart (778.84s) 
FAIL
FAIL    github.com/ethereum-optimism/optimism/espresso/devnet-tests     788.666s
FAIL
error: Recipe `devnet-enclave-tests` failed on line 47 with exit code 1
```
Started in tmux so I didn't fetch the full log TAT I can fetch the complete logs with a new run later

QuentinI committed:
WIP

philippecamacho commented:
How can we test this PR?

Did you try to run this test in CI by editing https://github.com/EspressoSystems/optimism-espresso-integration/blob/celo-integration-rebase-14.1/.github/workflows/espresso-enclave.yaml ?

cc @miguelCyclone

miguelCyclone committed:
WIP: TEE CI diagnostics

Ayiga committed:
Fix NON_TEE references

In the progress of this branch the constant `NON_TEE`
was replaced with a different constant.  In some
tests that were added while the work on this branch
were in progress, some new dependencies on `NON_TEE`
were added.  Over the course of rebases to address
these issues, this was reverted or otherwise missed.

This change fixes the missing constants.

dailinsubjam committed:
Fix lint and trigger ci

dailinsubjam committed:
Fix undefined NON_TEE constant reference

dailinsubjam committed:
not check linuxkit but check enclaver

dailinsubjam committed:
Skip TestSmokeWithTEE on non-enclave machines

dailinsubjam committed:
enable ESPRESSO_RUN_ENCLAVE_TESTS=true auto when just

dailinsubjam committed:
add TestSmokeWithTEE to CI

dailinsubjam committed:
add enclave-restart-test-on-ec2

dailinsubjam committed:
install docker v2

dailinsubjam committed:
create a succinct.env to make sure stop-containers can work

dailinsubjam committed:
Added set -a before sourcing .env

dailinsubjam committed:
start TestSmokeWithTEE's own CI job

shenkeyao commented on .github/workflows/espresso-devnet-tests.yaml:
```diff
@@ -24,7 +24,7 @@ jobs:
             tee: false
           - group: 3
-            tests: "TestSmokeWithTEE|TestForcedTransaction"
+            tests: "TestSmokeWithTEE"
```
Why do we remove `TestForcedTransaction`?

shenkeyao commented on espresso/devnet-tests/batcher_restart_test.go:
```diff
@@ -2,22 +2,48 @@ package devnet_tests
 
+	profile := DevnetProfileNonTee
+	if tee {
+		profile = DevnetProfileTee
```
Shouldn't we shut down and restart the TEE batcher if the profile is TEE? Looks like we are still only using the `op-batcher` service.

shenkeyao commented on espresso/devnet-tests/devnet_tools.go:
```diff
@@ -34,8 +34,24 @@ import (
 )
 
+// The setting for `COMPOSE_PROFILES` when running the Docker Compose.
+/*type DevnetProfile uint
+
+const (
+	DevnetProfileTee    DevnetProfile = iota
+	DevnetProfileNonTee DevnetProfile = iota
+)*/
```
Nit: can we delete this part?

shenkeyao commented on espresso/devnet-tests/devnet_tools.go:
This file could use some cleanups. There are many commented-out code snippets.

shenkeyao commented on op-batcher/enclave-tools/enclave-tools.go:
```diff
@@ -25,6 +26,11 @@ type EnclaveMeasurements struct {
 	PCR2 string `json:"PCR2"`
 }
 
+const errStrTxIdxingInProgress = "My Flag: transaction indexing is in progress"
```
Is this meant to be a temporary change, since it says "My Flag"?

shenkeyao commented on op-e2e/e2eutils/geth/wait.go:
```diff
@@ -62,6 +62,7 @@ func WaitForL1OriginOnL2(rollupCfg *rollup.Config, l1BlockNum uint64, client *et
 }
 
 func WaitForTransaction(hash common.Hash, client *ethclient.Client, timeout time.Duration) (*types.Receipt, error) {
+	//panic("FLAG A: Panic at WaitForTransaction")
```
Nit: remove this.

shenkeyao commented on espresso/environment/enclave_helpers.go:
```diff
@@ -41,7 +41,7 @@ const (
 )
 
-// Skips the calling test if `ESPRESSO_ENABLE_ENCLAVE_TESTS` is not set.
+// Skips the calling test if `ESPRESSO_RUN_ENCLAVE_TESTS` is not set.
```
The comment says `ESPRESSO_RUN_ENCLAVE_TESTS` but in the code it uses `ESPRESSO_ENABLE_ENCLAVE_TESTS`. There seems to be some naming inconsistency (worth double-checking other files too).
